src/tasks/worker.ts

```diff
@@ -0,0 +1,263 @@
+import type { AnyObject } from 'pinejs-client-core';
+import { tasks as tasksEnv } from '../config-loader/env';
+import type * as Db from '../database-layer/db';
+import { TransactionClosedError } from '../database-layer/db';
+import * as permissions from '../sbvr-api/permissions';
+import { PinejsClient } from '../sbvr-api/sbvr-utils';
+import { sbvrUtils } from '../server-glue/module';
+import { ajv, channel } from './common';
+import type { PartialTask, TaskHandler, TaskStatus } from './types';
+
+// Map of column names with SBVR names used in SELECT queries
+const selectColumns = Object.entries({
+	id: 'id',
+	'is executed by-handler': 'is_executed_by__handler',
+	'is executed with-parameter set': 'is_executed_with__parameter_set',
+	'is scheduled with-cron expression': 'is_scheduled_with__cron_expression',
+	'attempt count': 'attempt_count',
+	'attempt limit': 'attempt_limit',
+	priority: 'priority',
+	'is created by-actor': 'is_created_by__actor',
+})
+	.map(([key, value]) => `t."${key}" AS "${value}"`)
+	.join(', ');
+
+// The worker is responsible for executing tasks in the queue. It listens for
+// notifications and polls the database for tasks to execute. It will execute
+// tasks in parallel up to a certain concurrency limit.
+export class Worker {
+	public handlers: Record<string, TaskHandler> = {};
+	private readonly concurrency: number;
+	private readonly interval: number;
+	private client: PinejsClient;
+	private executing = 0;
+
+	constructor(client: PinejsClient) {
+		this.client = client;
+		this.concurrency = tasksEnv.queueConcurrency;
+		this.interval = tasksEnv.queueIntervalMS;
+	}
+
+	// Check if instance can execute more tasks
+	private canExecute(): boolean {
+		return (
+			this.executing < this.concurrency && Object.keys(this.handlers).length > 0
+		);
+	}
+
+	// Execute a task
+	private async execute(tx: Db.Tx, task: PartialTask): Promise<void> {
+		this.executing++;
+		try {
+			// Get specified handler
+			const handler = this.handlers[task.is_executed_by__handler];
+			const startedOnTime = new Date();
+			if (handler == null) {
+				await this.finalize(
+					tx,
+					task,
+					startedOnTime,
+					'failed',
+					'Matching task handler not found',
+				);
+				return;
+			}
+
+			// Validate parameters before execution so we can fail early if
+			// the parameter set is invalid. This can happen if the handler
+			// definition changes after a task is added to the queue.
+			if (
+				handler.validate != null &&
+				!handler.validate(task.is_executed_with__parameter_set)
+			) {
+				await this.finalize(
+					tx,
+					task,
+					startedOnTime,
+					'failed',
+					`Invalid parameter set: ${ajv.errorsText(handler.validate.errors)}`,
+				);
+				return;
+			}
+
+			// Execute handler
+			let status: TaskStatus = 'queued';
+			let error: string | undefined;
+			try {
+				await sbvrUtils.db.transaction(async (handlerTx) => {
```

Comment:
What's the reasoning between us handling the transaction creation? I could definitely imagine tasks that don't need a db transaction and so creating one for them would unnecessarily use up db connections

Reply:
@Page- The reasoning was probably to handle cases in which the transaction would be necessary, but I see your point and agree. We could accept a handler-level boolean option to determine if the task runner should create a transaction or not. Or just let the callers execute their logic within a transaction if they want and the task runner stays out of it. One reason I could see for creating transactions at the task runner level would be to reduce common duplicate code (a lot of tasks would want to execute within transactions).

Reply:
I would lean towards letting the task runner handle it and then if it's commonplace enough we can add sugar for making it easier. Basically I want solid foundations in the pinejs side that we can then extend/improve based on feedback from actual usage without preemptively adding solutions for problems that end up not existing (or needing a different solution), eg for this particular case it assumes a very narrow use case for tasks (ones that do db calls, and only db calls) and will work poorly for any that need to do stuff outside of a transaction (eg fetch some data in a transaction, close transaction, remote sync that data)